File: app/api/v1/schemas/conversations.py
"""会话相关 Schema。"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.api.v1.schemas.common import RequestIdMixin
from app.api.v1.schemas.rag import Citation


class ConversationListItem(BaseModel):
    """会话列表项。"""

    conversation_id: str = Field(description="会话ID")
    kb_id: str = Field(description="知识库ID")
    title: str | None = Field(default=None, description="会话标题")
    last_message_preview: str | None = Field(default=None, description="最后一条消息摘要")
    last_message_at: str | None = Field(default=None, description="最后一条消息时间")
    updated_at: str = Field(description="更新时间")


class ConversationListResponse(RequestIdMixin):
    """会话列表响应。"""

    items: list[ConversationListItem] = Field(description="会话列表")
    total: int = Field(default=0, description="会话总数")
    next_cursor: str | None = Field(default=None, description="下一页游标")


class ConversationCreateRequest(BaseModel):
    """创建会话请求。"""

    kb_id: str = Field(description="知识库ID")
    title: str | None = Field(default=None, description="会话标题")


class ConversationCreateResponse(RequestIdMixin):
    """创建会话响应。"""

    conversation_id: str = Field(description="会话ID")
    kb_id: str = Field(description="知识库ID")
    title: str | None = Field(default=None, description="会话标题")
    created_at: str = Field(description="创建时间")
    updated_at: str = Field(description="更新时间")


class ConversationRenameRequest(BaseModel):
    """重命名会话请求。"""

    title: str | None = Field(default=None, description="会话标题")


class ConversationDeleteResponse(RequestIdMixin):
    """删除会话响应。"""

    conversation_id: str = Field(description="会话ID")
    status: str = Field(description="删除状态")


class MessageItem(BaseModel):
    """消息条目。"""

    message_id: str = Field(description="消息ID")
    role: Literal["user", "assistant"] = Field(description="消息角色")
    content: str = Field(description="消息内容")
    citations: list[Citation] | None = Field(default=None, description="引用列表")
    refusal: bool | None = Field(default=None, description="是否拒答")
    refusal_reason: str | None = Field(default=None, description="拒答原因")
    timing: dict[str, int] | None = Field(default=None, description="耗时信息")
    created_at: str = Field(description="创建时间")


class ConversationDetailResponse(RequestIdMixin):
    """会话详情响应。"""

    conversation_id: str = Field(description="会话ID")
    kb_id: str = Field(description="知识库ID")
    messages: list[MessageItem] = Field(description="消息列表")


class MessagePageResponse(RequestIdMixin):
    """消息分页响应。"""

    items: list[MessageItem] = Field(description="消息列表")
    has_more: bool = Field(description="是否还有更多")
    next_before: str | None = Field(default=None, description="下一页 before 游标")

File: frontend/src/shared/api/modules/conversations.ts
import { apiClient } from "../client";
import { CitationItem, NextStepItem } from "./ask";

export interface ConversationListItem {
  conversation_id: string;
  kb_id: string;
  title?: string | null;
  last_message_preview?: string | null;
  last_message_at?: string | null;
  updated_at: string;
}

export interface ConversationListResponse {
  items: ConversationListItem[];
  total: number;
  next_cursor?: string | null;
  request_id?: string | null;
}

export interface ConversationMessage {
  message_id: string;
  parent_message_id?: string | null;
  edited_from_message_id?: string | null;
  role: "user" | "assistant";
  content: string;
  citations?: CitationItem[] | null;
  refusal?: boolean | null;
  refusal_reason?: string | null;
  suggestions?: string[] | null;
  next_steps?: NextStepItem[] | null;
  timing?: Record<string, number> | null;
  created_at: string;
  request_id?: string | null;
}

export interface ConversationDetailResponse {
  conversation_id: string;
  kb_id: string;
  messages: ConversationMessage[];
  request_id?: string | null;
}

export interface MessagePageResponse {
  items: ConversationMessage[];
  has_more: boolean;
  next_before?: string | null;
  request_id?: string | null;
}

export interface FeedbackPayload {
  rating: "up" | "down";
  reasons: string[];
  comment?: string;
  expected_hint?: string;
}

export interface ConversationCreatePayload {
  kb_id: string;
  title?: string;
}

export interface ConversationCreateResponse {
  conversation_id: string;
  kb_id: string;
  title?: string | null;
  created_at: string;
  updated_at: string;
  request_id?: string | null;
}

export interface ConversationRenamePayload {
  title?: string;
}

export interface ConversationListParams {
  kb_id?: string;
  keyword?: string;
  cursor?: string;
  limit?: number;
  offset?: number;
}

export interface ConversationMessagePageParams {
  before?: string;
  limit?: number;
}

export async function fetchConversationList(params?: {
  kb_id?: string;
  keyword?: string;
  cursor?: string;
  limit?: number;
  offset?: number;
}) {
  const { data } = await apiClient.get<ConversationListResponse>("/conversations", {
    params
  });
  return data;
}

export async function fetchConversationDetail(conversationId: string) {
  const { data } = await apiClient.get<ConversationDetailResponse>(
    `/conversations/${conversationId}`
  );
  return data;
}

export async function fetchConversationMessagesPage(
  conversationId: string,
  params?: ConversationMessagePageParams
) {
  const { data } = await apiClient.get<MessagePageResponse>(
    `/conversations/${conversationId}/messages`,
    { params }
  );
  return data;
}

export async function createConversation(payload: ConversationCreatePayload) {
  const { data } = await apiClient.post<ConversationCreateResponse>("/conversations", payload);
  return data;
}

export async function renameConversation(
  conversationId: string,
  payload: ConversationRenamePayload
) {
  const { data } = await apiClient.patch<ConversationCreateResponse>(
    `/conversations/${conversationId}`,
    payload
  );
  return data;
}

export async function deleteConversation(conversationId: string) {
  const { data } = await apiClient.delete<{
    conversation_id: string;
    status: "deleted";
    request_id?: string | null;
  }>(`/conversations/${conversationId}`);
  return data;
}

export async function submitFeedback(messageId: string, payload: FeedbackPayload) {
  const { data } = await apiClient.post<{
    feedback_id: string;
    message_id: string;
    status: string;
    request_id?: string | null;
  }>(`/messages/${messageId}/feedback`, payload);
  return data;
}
